controller/src/main/java/org/jboss/as/controller/AttributeMarshallers.java

```diff
@@ -55,12 +55,18 @@ public boolean isMarshallable(AttributeDefinition attribute, ModelNode resourceM
 
         @Override
         public void marshallAsElement(AttributeDefinition attribute, ModelNode resourceModel, boolean marshallDefault, XMLStreamWriter writer) throws XMLStreamException {
-            if (!resourceModel.hasDefined(attribute.getName())) {
+            if (!resourceModel.has(attribute.getName())) {
```

Comment:
For a properly implemented resource, this will always return 'true' because AbstractAddHandler.populateModel() will iterate over all attributes, for each adding a child (defined or undefined) to the resource's model.

I believe the behavior should be:

if (!resourceModel.hasDefined(attribute.getName())) -> return without marshalling anything for the attribute
List<ModelNode> list = resource.get(attribute.getName()).asList();
if (list.isEmpty()) {
   if (wrapElement) -> marshal an empty wrapper
   else -> return without marshalling anything (see below)
}
else -> [marshal a start element] / iterate over list / [marshal end element].

A problem here is "if (list.isEmpty() && !wrapElement)". This indicates a subtle programming error, where the developer doesn't provide an xml schema that prevents ambiguity between the 'null' and 'empty collection' cases. I don't think we should try to deal with this at this stage of the release process though, other than perhaps adding a TODO and filing a JIRA. A fix would likely involve throwing an ISE or AssertionError in at least some cases where ^^^ is true, and that's too risky right now.